src/service/AgentCertificateService.ts

```diff
@@ -53,20 +95,74 @@ export class AgentCertificateService {
+        }
+        const agentCertificate = certificates[0];
+        BootstrapUtils.validateIsTrue(agentCertificate.privateKey === agentPrivateKey, 'Invalid agent private key');
+        BootstrapUtils.validateIsTrue(agentCertificate.publicKey === providedCertificates.agent.publicKey, 'Invalid agent public key');
         logger.info(`Agent Certificate for node ${name} created`);
+        const metadata: AgentCertificateMetadata = {
+            version: AgentCertificateService.METADATA_VERSION,
+            agentPublicKey: providedCertificates.agent.publicKey,
+        };
+        await BootstrapUtils.writeYaml(metadataFile, metadata, undefined);
+    }
+    private async shouldGenerateCertificate(metadataFile: string, providedCertificates: AgentCertificates): Promise<boolean> {
+        if (!existsSync(metadataFile)) {
+            return true;
+        }
+        try {
+            const metadata = BootstrapUtils.loadYaml(metadataFile, false) as AgentCertificateMetadata;
+            return (
+                metadata.agentPublicKey !== providedCertificates.agent.publicKey ||
+                metadata.version !== AgentCertificateService.METADATA_VERSION
+            );
+        } catch (e) {
+            logger.warn(`Cannot load agent certificate metadata from file ${metadataFile}. Error: ${e.message}`, e);
+            return true;
+        }
     }
 
+    // Alternative reading the public key from agent-ca.pubkey.pem
+    // It's ok but it requires sshpk and unsure how to force an regeneration
+    // private async shouldGenerateCertificate(certFolder: string, providedCertificates: AgentCertificates): Promise<boolean> {
+    //   const pemFile = join(certFolder, 'agent-ca.pubkey.pem');
+    //   if (!existsSync(pemFile)) {
+    //     return true;
+    //   }
+    //   try {
+    //     const data = readFileSync(pemFile);
+    //     const key = sshpk.parseKey(data, 'pem');
+    //     const storedPublicKey = Convert.uint8ToHex((key.part as any).A.data);
+    //     return storedPublicKey !== providedCertificates.agent.publicKey;
+    //   } catch (e) {
+    //     logger.warn(`Cannot extract public key from file ${pemFile}. Error: ${e.message}`, e);
+    //     return true;
+    //   }
+    // }
+
     private createCertCommands(target: string): string {
         return `set -e
-        cd ${target}
+cd ${target}
+
+# prepare dirs/files
+chmod 700 new_certs
+touch index.txt
+
 # Creating a self signed certificate for the agent. This will be created by bootstrap when setting up a supernode
-openssl genrsa -out agent-key.pem 4096
-openssl req -new -config agent.cnf -key agent-key.pem -out agent-csr.pem
-openssl x509 -req -extfile agent.cnf -days 999 -in agent-csr.pem -out agent-crt.pem -signkey agent-key.pem
+
+cat agent-ca.der | openssl pkey -inform DER -outform PEM -out agent-ca.key.pem
+openssl pkey -inform pem -in agent-ca.key.pem -text -noout
```

Comment:
just to make sure, `agent-ca.der` is generated using `ed25519` algorithm, right?

Reply:
Yes, it also check the openssl private and public keys against the sdk generated ones